### PR TITLE
feat: add DataFrame.pipe() method

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: tests/system/small/test_dataframe.py [hunk 1]
```diff
@@ -983,6 +983,31 @@ def test_apply_series_scalar_callable(
     pandas.testing.assert_series_equal(bf_result, pd_result)
 
 
+def test_df_pipe(
```
Comment (shobsi) [inline tests/system/small/test_dataframe.py:986]: let's add a series test too

File: third_party/bigframes_vendored/pandas/core/common.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    obj, func: Callable[..., T] | tuple[Callable[..., T], str], *args, **kwargs
+) -> T:
+    """
+    Apply a function ``func`` to object ``obj`` either by passing obj as the
+    first argument to the function or, in the case that the func is a tuple,
+    interpret the first element of the tuple as a function and pass the obj to
+    that function as a keyword argument whose key is the value of the second
+    element of the tuple.
+
+    Parameters
```
Comment (shobsi) [inline third_party/bigframes_vendored/pandas/core/common.py:20]: google style docstring please.

Also could you please screenshots of the API doc rendering?

File: third_party/bigframes_vendored/pandas/core/common.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        object.
+    *args : iterable, optional
+        Positional arguments passed into ``func``.
+    **kwargs : dict, optional
+        A dictionary of keyword arguments passed into ``func``.
+
+    Returns
+    -------
+    object : the return type of ``func``.
+    """
```
Comment (shobsi) [inline third_party/bigframes_vendored/pandas/core/common.py:35]: Code samples please. We could add same examples as pandas https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pipe.html

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 2 above)
Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/common.py:20]: Fixed to follow google style docstrings.

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 3 above)
Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/common.py:35]: copied all examples from pandas docstring.

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 2 above)
Comment (shobsi) [reply third_party/bigframes_vendored/pandas/core/common.py:20]: Examples are not rendering (they go before Args/Returns). also the args and kwargs appear funky https://screenshot.googleplex.com/97cktrgpWUyMzxL

Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/common.py:20]: fixed docstring. haven't got the rendering process figure out yet though

File: third_party/bigframes_vendored/pandas/core/generic.py [hunk 4]
```diff
@@ -962,6 +966,105 @@ def expanding(self, min_periods=1):
...
+        Constructing a income DataFrame from a dictionary.
+
+            >>> import bigframes.pandas as bpd
+            >>> import numpy as np
+            >>> bpd.options.display.progress_bar = None
+
+            >>> data = [[8000, 1000], [9500, np.nan], [5000, 2000]]
+            >>> df = bpd.DataFrame(data, columns=['Salary', 'Others'])
+            >>> df
+            Salary  Others
```
Comment (shobsi) [inline third_party/bigframes_vendored/pandas/core/generic.py:989]: Indentation https://screenshot.googleplex.com/9SMdYmvkeop6gua

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 2 above)
Comment (shobsi) [reply third_party/bigframes_vendored/pandas/core/common.py:20]: Thanks! [nit] The style in this file (common.py) itself is not google style, but it isn't leading to the website as the one in generic.py is taking effect, so we can remove it from here, or update to google style for consistency.

File: third_party/bigframes_vendored/pandas/core/common.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Contains code from https://github.com/pandas-dev/pandas/blob/main/pandas/core/common.py
+from __future__ import annotations
+
+from typing import Callable, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from third_party.bigframes_vendored.pandas.pandas._typing import T
```
Comment (tswast) [inline third_party/bigframes_vendored/pandas/core/common.py:7]: This needs to be synced with `main` and updated.

```suggestion
    from bigframes_vendored.pandas.pandas._typing import T
```

File: third_party/bigframes_vendored/pandas/core/generic.py [hunk 6]
```diff
@@ -1,10 +1,14 @@
 # Contains code from https://github.com/pandas-dev/pandas/blob/main/pandas/core/generic.py
 from __future__ import annotations
 
-from typing import Iterator, Literal, Optional
+from typing import Callable, Iterator, Literal, Optional, TYPE_CHECKING
 
 from bigframes import constants
 from third_party.bigframes_vendored.pandas.core import indexing
+import third_party.bigframes_vendored.pandas.core.common as common
```
Comment (tswast) [inline third_party/bigframes_vendored/pandas/core/generic.py:8]: Likewise, needs a sync with `main` and update.

```suggestion
import bigframes_vendored.pandas.core.common as common
```

Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/generic.py]: done

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 5 above)
Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/common.py]: done

File: third_party/bigframes_vendored/pandas/core/generic.py (hunk 4 above)
Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/generic.py]: done

File: third_party/bigframes_vendored/pandas/core/common.py (hunk 2 above)
Comment (TrevorBergeron) [reply third_party/bigframes_vendored/pandas/core/common.py]: sorry, forgot that I had two docstrings to fix. changed this one to google style too

**Review by shobsi — APPROVED** [review]

**Pull request merged**